File: V0.2/linkguardMB/Android/app/src/main/java/com/linkguard/app/engine/SimulationEngine.kt
package com.linkguard.app.engine

import com.linkguard.app.model.*

class SimulationEngine {

    private val victimIDs = listOf("VT-A3F", "VT-B72", "VT-C1E", "VT-D9A", "VT-E45")

    fun createInitialNodeStatus(): RescueNodeStatus {
        return RescueNodeStatus(
            nodeID = "RT-7F2-EMT",
            deptCode = "EMT",
            battery = 85,
            loraLevel = 4,
            isConnected = false,
            pairCode = "0000"
        )
    }

    fun createInitialVictims(): List<VictimNode> {
        val now = System.currentTimeMillis()
        return listOf(
            VictimNode("VT-A3F", 78, 65, -62.0, 8.5, true, now, true),
            VictimNode("VT-B72", 92, 43, -78.0, 5.2, false, now - 10000, true),
            VictimNode("VT-C1E", 0, 12, -91.0, 1.3, true, now - 45000, false),
        )
    }

    fun createInitialSOSRecords(): List<SOSRecord> {
        val now = System.currentTimeMillis()
        return listOf(
            SOSRecord(victimID = "VT-A3F", heartRate = 78, rssi = -62.0,
                distance = "1.2m", battery = 65, time = now - 30000, isAcknowledged = false),
            SOSRecord(victimID = "VT-C1E", heartRate = 0, rssi = -91.0,
                distance = "48m", battery = 12, time = now - 120000, isAcknowledged = true),
        )
    }

    fun updateVictimSignals(victims: MutableList<VictimNode>) {
        for (i in victims.indices) {
            if (!victims[i].isOnline) continue
            val rssiDrift = (-3..3).random().toDouble()
            val snrDrift = (-0.5 + Math.random()).let { it - 0.5 }
            val hrDrift = (-3..3).random()

            victims[i] = victims[i].copy(
                rssi = (victims[i].rssi + rssiDrift).coerceIn(-99.0, -30.0),
                snr = (victims[i].snr + snrDrift).coerceIn(-5.0, 15.0),
                heartRate = if (victims[i].heartRate > 0)
                    (victims[i].heartRate + hrDrift).coerceIn(40, 180) else 0,
                lastSeen = System.currentTimeMillis(),
                battery = if ((0..15).random() == 0)
                    maxOf(0, victims[i].battery - 1) else victims[i].battery
            )
        }
    }

    fun updateVictimOnlineStatus(victims: MutableList<VictimNode>) {
        val now = System.currentTimeMillis()
        for (i in victims.indices) {
            val interval = (now - victims[i].lastSeen) / 1000
            var online = interval < 30
            if (!online && (0..10).random() == 0) {
                online = true
                victims[i] = victims[i].copy(
                    isOnline = true,
                    rssi = (-85..-50).random().toDouble(),
                    lastSeen = now
                )
            } else {
                victims[i] = victims[i].copy(isOnline = online)
            }
        }
    }

    fun updateNodeStatus(status: RescueNodeStatus): RescueNodeStatus {
        return if ((0..30).random() == 0) {
            status.copy(battery = maxOf(0, status.battery - 1))
        } else status
    }

    fun maybeDiscoverNewVictim(existing: List<VictimNode>): VictimNode? {
        if ((0..30).random() != 0) return null
        val existingIDs = existing.map { it.id }.toSet()
        val available = victimIDs.filter { it !in existingIDs }
        val newID = available.randomOrNull() ?: return null
        return VictimNode(
            id = newID,
            heartRate = if ((0..1).random() == 0) (50..120).random() else 0,
            battery = (10..90).random(),
            rssi = (-90..-50).random().toDouble(),
            snr = (0..10).random().toDouble(),
            isSOS = (0..2).random() == 0,
            lastSeen = System.currentTimeMillis(),
            isOnline = true
        )
    }

    fun maybeToggleSOS(victims: MutableList<VictimNode>): SOSRecord? {
        if (victims.isEmpty()) return null
        val i = victims.indices.random()
        if (!victims[i].isOnline || (0..20).random() != 0) return null
        victims[i] = victims[i].copy(isSOS = !victims[i].isSOS)
        if (!victims[i].isSOS) return null
        return SOSRecord(
            victimID = victims[i].id,
            heartRate = victims[i].heartRate,
            rssi = victims[i].rssi,
            distance = victims[i].distanceText,
            battery = victims[i].battery,
            time = System.currentTimeMillis()
        )
    }
}

File: v0.4_Rebuild_INSARAG/apps/ios/LinkGuard-TL/LinkGuard-TL/V02FieldApp/SimulationEngine.swift
import Foundation

/// 模擬引擎：模擬韌體 rescue.ino 的行為，產生擬真受困者資料
/// 用於 Demo 展示（沒有硬體時可模擬 BLE 資料）
class SimulationEngine {

    private let victimIDs = ["VT-A3F", "VT-B72", "VT-C1E", "VT-D9A", "VT-E45"]

    // MARK: - 初始資料

    func createInitialNodeStatus() -> RescueNodeStatus {
        RescueNodeStatus(
            nodeID: "RT-7F2-EMT",
            deptCode: "EMT",
            battery: 85,
            loraLevel: 4,
            isConnected: false,
            pairCode: "0000"
        )
    }

    func createInitialVictims() -> [VictimNode] {
        [
            VictimNode(id: "VT-A3F", heartRate: 78, battery: 65, rssi: -62, snr: 8.5,
                       isSOS: true, lastSeen: Date(), isOnline: true),
            VictimNode(id: "VT-B72", heartRate: 92, battery: 43, rssi: -78, snr: 5.2,
                       isSOS: false, lastSeen: Date().addingTimeInterval(-10), isOnline: true),
            VictimNode(id: "VT-C1E", heartRate: 0, battery: 12, rssi: -91, snr: 1.3,
                       isSOS: true, lastSeen: Date().addingTimeInterval(-45), isOnline: false),
        ]
    }

    func createInitialSOSRecords() -> [SOSRecord] {
        let now = Date()
        return [
            SOSRecord(id: UUID(), victimID: "VT-A3F", heartRate: 78, rssi: -62,
                      distance: "1.2m", battery: 65, time: now.addingTimeInterval(-30), isAcknowledged: false),
            SOSRecord(id: UUID(), victimID: "VT-C1E", heartRate: 0, rssi: -91,
                      distance: "48m", battery: 12, time: now.addingTimeInterval(-120), isAcknowledged: true),
        ]
    }

    // MARK: - 模擬更新（每 2 秒 tick，對應韌體 PING_INTERVAL=3000）

    /// 模擬 RSSI 訊號波動與心率微變
    func updateVictimSignals(_ victims: inout [VictimNode]) {
        for i in victims.indices where victims[i].isOnline {
            // RSSI 微幅波動
            let rssiDrift = Double.random(in: -3...3)
            victims[i].rssi = max(-99, min(-30, victims[i].rssi + rssiDrift))
            // SNR 微幅波動
            let snrDrift = Double.random(in: -0.5...0.5)
            victims[i].snr = max(-5, min(15, victims[i].snr + snrDrift))
            // 心率微變
            if victims[i].heartRate > 0 {
                let hrDrift = Int.random(in: -3...3)
                victims[i].heartRate = max(40, min(180, victims[i].heartRate + hrDrift))
            }
            victims[i].lastSeen = Date()
            // 偶爾消耗電量
            if Int.random(in: 0...15) == 0 {
                victims[i].battery = max(0, victims[i].battery - 1)
            }
        }
    }

    /// 模擬受困者上線/離線（韌體：30s 無更新 = 離線）
    func updateVictimOnlineStatus(_ victims: inout [VictimNode]) {
        for i in victims.indices {
            let interval = Date().timeIntervalSince(victims[i].lastSeen)
            victims[i].isOnline = interval < 30
            // 偶爾讓離線的重新上線
            if !victims[i].isOnline && Int.random(in: 0...10) == 0 {
                victims[i].isOnline = true
                victims[i].rssi = Double.random(in: -85 ... -50)
                victims[i].lastSeen = Date()
            }
        }
    }

    /// 模擬節點電量與 LoRa 檔位
    func updateNodeStatus(_ status: inout RescueNodeStatus) {
        if Int.random(in: 0...30) == 0 {
            status.battery = max(0, status.battery - 1)
        }
    }

    /// 隨機產生新受困者 SOS（低機率，模擬 LoRa 收到新封包）
    func maybeDiscoverNewVictim(existing: [VictimNode]) -> VictimNode? {
        guard Int.random(in: 0...30) == 0 else { return nil }
        // 從 pool 中找一個尚未存在的 ID
        let existingIDs = Set(existing.map(\.id))
        let available = victimIDs.filter { !existingIDs.contains($0) }
        guard let newID = available.randomElement() else { return nil }
        return VictimNode(
            id: newID,
            heartRate: Int.random(in: 0...1) == 0 ? Int.random(in: 50...120) : 0,
            battery: Int.random(in: 10...90),
            rssi: Double.random(in: -90 ... -50),
            snr: Double.random(in: 0...10),
            isSOS: Int.random(in: 0...2) == 0,
            lastSeen: Date(),
            isOnline: true
        )
    }

    /// 隨機觸發 SOS 狀態變化
    func maybeToggleSOS(_ victims: inout [VictimNode]) -> SOSRecord? {
        guard let i = victims.indices.randomElement(),
              victims[i].isOnline,
              Int.random(in: 0...20) == 0
        else { return nil }

        victims[i].isSOS.toggle()

        guard victims[i].isSOS else { return nil }
        return SOSRecord(
            id: UUID(),
            victimID: victims[i].id,
            heartRate: victims[i].heartRate,
            rssi: victims[i].rssi,
            distance: victims[i].distanceText,
            battery: victims[i].battery,
            time: Date(),
            isAcknowledged: false
        )
    }
}
